### DISCUSSION TITLE
stale option doesn't work in refetch method

### DISCUSSION THREAD

**Discussion by fhugoduarte** [TanStack/query#4432] [Q&A] [answered]
#### Descrption 

I would like to call my query imperatively. For that I passed the **enable** option as **false** and I'm using the **refetch** to execute the query.

The problem is that I want the **refetch** method to respect the **staleTime**, I'm trying to pass the **stale option as true**, but the request is fired anyway.

I would like to know if this is the expected behavior and how I could solve this problem.
_Currently I'm passing enabled as a parameter to my query and controlling it with a state._

#### Evidence: 

https://user-images.githubusercontent.com/26551306/199591186-debb3cb0-25b9-4535-9fa2-2318beb88729.mov

[CodeSandBox example](https://codesandbox.io/s/hungry-euler-yz1xr5?file=/src/App.tsx)

**Comment by TkDodo** [comment, marked as answer]
Yes, this is the expected behaviour. `refetch` returned from `useQuery` is an unconditional fetch - the only thing that can bypass the `enabled` option. By doing this, you opt-out of most of the cool features that react-query offers, like respecting `staleTime`.

> I would like to call my query imperatively. For that I passed the enable option as false and I'm using the refetch to execute the query.

I would argue that this is still not the right way to go about it in most cases. You can have queries be lazily evaluated with a dynamic `enabled` option. Have a look at [lazy queries](https://tanstack.com/query/v4/docs/guides/disabling-queries#lazy-queries).

**Reply by GeorgeCht** [reply]
If anyone is still searching this [check here](https://tanstack.com/query/latest/docs/framework/react/guides/disabling-queries#lazy-queries)

**Comment by fhugoduarte** [comment]
Ok, but the refetch function takes these parameters. I still don't understand why they exist, do they really work? 
Wouldn't it be possible to add these kind of filters to the refetch?

**Reply by TkDodo** [reply]
> Ok, but the refetch function takes these parameters. I still don't understand why they exist, do they really work?


not sure what you are talking about. `refetch` doesn't take any parameters. Are we talking about the same thing?

```
const { refetch } = useQuery(...)

// ⬇️ no parameters here!
refetch()
```

**Reply by fhugoduarte** [reply]
Does refetch not receive options params? [CodeSandBox example](https://codesandbox.io/s/hungry-euler-yz1xr5?file=/src/App.tsx)

<img width="698" alt="Screen Shot 2022-11-27 at 13 05 26" src="https://user-images.githubusercontent.com/26551306/204145038-e572d43a-fc88-46b1-9df4-72bce98e2400.png">

**Reply by TkDodo** [reply]
no, it takes:

```
throwOnError?: boolean
cancelRefetch?: boolean
refetchPage?: (
  lastPage: TPageData,
  index: number,
  allPages: TPageData[],
) => boolean
```

https://github.com/TanStack/query/blob/1614c31f84744c62b5bee8f2ff97fc15a1652e86/packages/query-core/src/queryObserver.ts#L277-L280

all of these are "options" that you'll rarely need, not filters.

The interesting thing is that in your sandbox, the type for refetch is shown as:

`RefetchOptions & RefetchQueryFilters<unknown>`

and the `RefetchQueryFilters` part is certainly wrong, so I think we have a typings bug here:

https://github.com/TanStack/query/blob/07c96311a50a2aa45f516fa2c153ce68ca8f1a9b/packages/query-core/src/types.ts#L389-L391

It should be `RefetchPageFilters`, not `RefetchQueryFilters` 😅 . Feel free to open an issue, or even better, a PR :)
